include/AudioPluginInterface.h

```diff
@@ -0,0 +1,418 @@
+	 */
+	//virtual void processImpl(NotePlayHandle* nph) {}
+
+	/**
+	 * The main audio processing method for MIDI-based Instruments.
+	 * The implementation knows how to provide the working buffers.
+	 */
+	virtual void processImpl() = 0;
+};
+
+//! Effect specialization
+template<typename BufferT, typename ConstBufferT>
+class AudioProcessingMethod<Effect, BufferT, ConstBufferT, false, false>
+{
+protected:
+	/**
+	 * The main audio processing method for Effects. Runs when plugin is not asleep.
+	 * The implementation is expected to perform wet/dry mixing for the first 2 channels.
+	 */
+	virtual auto processImpl(ConstBufferT in, BufferT out) -> ProcessStatus = 0;
+};
+
+//! Effect specialization (in-place)
+template<typename BufferT, typename ConstBufferT>
+class AudioProcessingMethod<Effect, BufferT, ConstBufferT, true, false>
+{
+protected:
+	/**
+	 * The main audio processing method for inplace Effects. Runs when plugin is not asleep.
+	 * The implementation is expected to perform wet/dry mixing for the first 2 channels.
+	 */
+	virtual auto processImpl(BufferT inOut) -> ProcessStatus = 0;
+};
+
+//! Effect specialization (custom working buffers)
+template<typename BufferT, typename ConstBufferT, bool inplace>
+class AudioProcessingMethod<Effect, BufferT, ConstBufferT, inplace, true>
+{
+protected:
+	/**
+	 * The main audio processing method for Effects. Runs when plugin is not asleep.
+	 * The implementation knows how to provide the working buffers.
+	 * The implementation is expected to perform wet/dry mixing for the first 2 channels.
+	 */
+	virtual auto processImpl() -> ProcessStatus = 0;
+};
+
+//! Connects the core audio buses to the instrument or effect using the pin connector
+template<class ChildT, typename SampleT, PluginConfig config>
+class AudioProcessorImpl
+{
+	static_assert(always_false_v<ChildT>, "ChildT must be either Instrument or Effect");
+};
+
+//! Instrument specialization
+template<typename SampleT, PluginConfig config>
+class AudioProcessorImpl<Instrument, SampleT, config>
+	: public Instrument
+	, public AudioProcessingMethod<Instrument,
+		typename AudioDataTypeSelector<config.layout, SampleT, config.inputs>::type,
+		typename AudioDataTypeSelector<config.layout, const SampleT, config.outputs>::type,
+		config.inplace, config.customBuffer>
+	, public std::conditional_t<config.customBuffer,
+		AudioPluginBufferInterfaceProvider<config.layout, SampleT, config.inputs, config.outputs>,
+		AudioPluginBufferDefaultImpl<config.layout, SampleT, config.inputs, config.outputs, config.inplace>>
+{
+public:
+	AudioProcessorImpl(const Plugin::Descriptor* desc, InstrumentTrack* parent = nullptr, const Plugin::Descriptor::SubPluginFeatures::Key* key = nullptr, Instrument::Flags flags = Instrument::Flag::NoFlags)
+		: Instrument{desc, parent, key, flags}
+		, m_pinConnector{config.inputs, config.outputs, this}
+	{
+		connect(Engine::audioEngine(), &AudioEngine::sampleRateChanged, [this]() {
```

Comment:
I'm considering maybe making `PluginPinConnector` the parent class of the audio plugin buffer class which would remove the need for a bunch of `connect` calls. Currently the pin connector and the plugin's buffer class do not interact with each other directly but instead rely on Qt signals and `AudioPluginInterface` to set up those connections, which may be complicating things.

For example, the plugin buffer initialization for VST plugins is awkward. Currently in `RemotePlugin`, it has to call the pin connector's method to update the channel counts then follow that with an `updateBuffers()` call. Normally calling the pin connector's method to update the channel counts would emit a signal that tells the buffer it needs to update, but this doesn't work while a VST plugin is being initialized (since `m_plugin` is null until initialization is complete, but it needs to be non-null for `AudioPluginBuffer` to access the plugin's buffer to tell it to update itself).

Making the pin connector the base class of the plugin buffer, or coupling the two classes directly together by some other means could solve this issue and remove the need for the `connect` calls.